### PR TITLE
[TorchOnnxToTorch] Numerically stabilize ONNX Softplus lowering

### PR DESCRIPTION
**Pull request by nmalimban**
## Summary

Numerically stabilizes the `TorchOnnxToTorch` lowering of ONNX `Softplus`.

The conversion previously emitted the naive form `ln(exp(x) + 1)`, which overflows float32 to `+inf` once `x` exceeds ~88 (`exp(88) ≈ 1.65e38`, near the fp32 max `3.4e38`). A single overflowing activation poisons every downstream value with `+inf`, diverging from the numerically stable form used by ONNX Runtime and other importers.

This PR rewrites the lowering to the standard stable identity:

```
softplus(x) = ln(exp(x) + 1) = max(x, 0) + log1p(exp(-|x|))
```

which keeps the exponent's argument `<= 0`, so `exp(-|x|)` stays in `(0, 1]` and never overflows.

## Details: derivation of the identity

Let `m = max(x, 0)`. Factor `e^m` out of the sum inside the logarithm:

```
softplus(x) = ln(e^x + 1)
            = ln(e^x + e^0)
            = ln( e^m * (e^(x-m) + e^(-m)) )
            = m + ln( e^(x-m) + e^(-m) )
```

Because `m = max(x, 0)`, one of the two exponents `x-m`, `-m` is exactly `0` and the other is `<= 0`:

- if `x >= 0`, then `m = x`, so `x - m = 0` and `-m = -x = -|x|`;
- if `x <  0`, then `m = 0`, so `-m = 0` and `x - m = x = -|x|`.

In both cases `{e^(x-m), e^(-m)} = {1, e^(-|x|)}`, so the sum is `1 + e^(-|x|)` and

```
softplus(x) = m + ln(1 + e^(-|x|)) = max(x, 0) + log1p(exp(-|x|)).
```

`log1p` is used rather than `log(1 + x)` for accuracy when `e^(-|x|)` is small (i.e. large `|x|`).

**Why this is stable.** The only exponentiation is of `-|x| <= 0`, so `exp(-|x|) in (0, 1]` and `1 + exp(-|x|) in (1, 2]` — both always finite in float32 regardless of the magnitude of `x`. The unbounded term `max(x, 0)` only ever appears as a plain add, never inside an exponential, so it cannot overflow. The naive form, by contrast, exponentiates `x` directly and overflows once `x` exceeds ~88 in float32.

## Changes

- **`DefaultDomainQtoZ.cpp`** — the ONNX `Softplus` binder now emits `relu(x) + log1p(exp(-|x|))` (using `relu(x) = max(x, 0)`) instead of `log1p(exp(x))`.
- **`simple_ops_q_to_z.mlir`** — FileCheck coverage updated to match the new op sequence (`abs → neg → exp → log1p → relu → add`).

## Testing

Verified manually:

- **lit / FileCheck**: `test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir` passes.
- Matches `torch.nn.functional.softplus` exactly (max abs diff `0.0` over `x` in `[-50, 200]`, including `x = 109` where the naive form returns `inf`), and lowers cleanly to linalg and TOSA.

### CODE REVIEW

**Comment by nmalimban** [comment]
Hi, @sahas3 @Lallapallooza @rengolin. I am requesting reviews via this comment since I do not yet have Triage permissions for this repo. Thanks!

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 1]
```diff
@@ -2618,11 +2618,26 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
             binder.tensorResultType(resultType)) {
           return failure();
         }
-        // out = ln(exp(x) + 1)
-        Value exp = Torch::AtenExpOp::create(rewriter, binder.getLoc(),
-                                             resultType, input);
-        rewriter.replaceOpWithNewOp<Torch::AtenLog1pOp>(binder.op, resultType,
-                                                        exp);
+        Location loc = binder.getLoc();
```
Comment (sahas3) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2621]: Can you add an e2e test with input > 88 to lockdown the numeric stability?

**Review by sahas3 — COMMENTED** [review]
Thanks for the patch, looks good modulo the comment.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp:2621]: Originally, I had intended to add an e2e test, but I discovered that an e2e ONNX test does not actually catch this.

The reason for this is that the --config=onnx e2e suite always goes through torch.onnx.export. Every test is registered as a torch.nn.Module and it is exported via torch.onnx.export(..., 
  dynamo=True) before torch-mlir sees anything. That exporter faithfully reproduces PyTorch's softplus(beta=1, threshold=20) semantics, so it doesn't emit a bare Softplus — it emits:

  graph nodes: ['Softplus', 'Greater', 'Where']   # Where(x > 20, x, Softplus(x))
 
For any x > 20 the graph returns x directly and the Softplus branch is masked out.

So the overflow issue issue can only be hit if something produces an ONNX softplus without going through PyTorch's exporter.

Do you know if there is a good way to write an e2e test, given this? I decided not to pursue this, especially since the LIT test already locks down the numerically-stable operation graph.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 2]
```diff
@@ -2618,11 +2618,26 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+        // Softplus(x) = ln(exp(x) + 1), computed in the numerically stable form
+        //   max(x, 0) + log1p(exp(-|x|))
+        // to avoid overflowing exp(x) to +inf (in fp32, exp overflows once
+        // x exceeds ~88). The naive ln(exp(x) + 1) poisons every downstream
+        // value with +inf for large activations; factoring out max(x, 0) keeps
+        // the exponent's argument <= 0, so exp(-|x|) stays in (0, 1].
+        Value absX = Torch::AtenAbsOp::create(rewriter, loc, resultType, input);
```
Comment (hariprasadravi) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Could the binder just emit torch.aten.softplus? We already have the op (GeneratedTorchOps.td:5502) with a decomposition (DecomposeComplexOps.cpp:7383) that's addIllegalOp in LowerToBackendContract.cpp:495, so it always decomposes and no backend needs its own lowering. We will emit one op instead of six, and it keeps the ONNX and FX paths on the same numerics. I checked that decomposition against eager over the same range including the overflow region.

Though I'd argue your identity is better than what that decomposition does.
It uses PyTorch's where(x*beta > threshold, x, log1p(exp(x*beta))/beta), and
since where evaluates both arms the exp still overflows to inf -- the select
just discards it. Your form never materializes an inf, and it matches the ONNX
spec, which is exact ln(e^x + 1) with no threshold. So my mild preference would
be the inverse of this patch: move your identity into DecomposeAtenSoftplusOp
and have the binder emit aten.softplus. That fixes the FX/TorchScript path too.

 Either direction is fine by me. My only real concern is that as written, the
 same mathematical function now lowers to two different op sequences depending
 on the frontend.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 3]
```diff
@@ -2618,11 +2618,26 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+        Value expNegAbsX =
+            Torch::AtenExpOp::create(rewriter, loc, resultType, negAbsX);
+        Value log1p =
+            Torch::AtenLog1pOp::create(rewriter, loc, resultType, expNegAbsX);
+        Value reluX =
+            Torch::AtenReluOp::create(rewriter, loc, resultType, input);
+        Value one = Torch::ConstantIntOp::create(rewriter, loc,
+                                                 rewriter.getI64IntegerAttr(1));
+        rewriter.replaceOpWithNewOp<Torch::AtenAddTensorOp>(
+            binder.op, resultType, reluX, log1p, one);
```
Comment (hariprasadravi) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: The one thew me off. Consider adding a \/*alpha=\*/

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 2 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: This makes sense to me, I can make this change. One note, though, is that `aten.softplus` exposes the `Threshold` as a parameter (in PyTorch, seems like you call `torch.nn.Softplus(beta=1.0, threshold=20.0)`). So if I were to make this change in `DecomposeAtenSoftplusOp`, I would still implement the threshold branch, but fix the decomp for the branch beneath the threshold. This means, from a numeric-accuracy perspective, there is no real reason to set the threshold, since the branch below the threshold will be numerically stable. I think this is fine.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 4]
```diff
@@ -2610,21 +2610,44 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+                  // numerically stable form max(beta*x, 0) +
+                  // log1p(exp(-|beta*x|)) rather than the overflow-prone
+                  // ln(exp(x) + 1). Routing through the shared op keeps the
+                  // ONNX, FX, and TorchScript frontends on identical numerics.
+                  //
+                  // beta = 1 matches the ONNX spec. threshold = 20 is PyTorch's
+                  // default: for beta*x > threshold aten.softplus returns x
+                  // directly instead of the full formula. That shortcut is
+                  // bit-exact against ONNX in fp32, because what it drops is
+                  // below the fp32 rounding granularity at x = 20:
```
Comment (hariprasadravi) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: e2e test for x >= 20 will be nice

**Review by hariprasadravi — COMMENTED** [review]
LG

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp [hunk 5]
```diff
@@ -2610,21 +2610,44 @@ void mlir::torch::onnx_c::populateDefaultDomainQtoZ(
...
+  patterns.onOp("Softplus", 1,
+                [](OpBinder binder, ConversionPatternRewriter &rewriter) {
+                  Torch::ValueTensorType resultType;
+                  Value input;
+                  if (binder.tensorOperand(input) ||
+                      binder.tensorResultType(resultType)) {
+                    return failure();
+                  }
+                  Location loc = binder.getLoc();
+                  // ONNX Softplus(x) = ln(exp(x) + 1). Emit aten.softplus (beta
```
Comment (sahas3) [inline lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Do we really need to capture this commentary in the code? It seems quite natural that `onnx.softplus` is mapped to `aten.softplus` as part of converting ONNX IR to torch. Just mentioning the choice of `beta=1, threshold=20` is sufficient.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py [hunk 6]
```diff
@@ -2389,6 +2389,35 @@ def SoftplusModule_basic(module, tu: TestUtils):
...
+    )
+    def forward(self, x):
+        return torch.ops.aten.softplus(x)
+
+
+@register_test_case(module_factory=lambda: SoftplusLargeMagnitudeModule())
+def SoftplusLargeMagnitudeModule_basic(module, tu: TestUtils):
+    # Inputs on both sides of the fp32 exp overflow threshold (~88). The stable
+    # (max(z, 0) + log1p(exp(-|z|))) / beta arm in DecomposeAtenSoftplusOp keeps
+    # these finite and matches eager, where the old naive log1p(exp(z)) arm
```
Comment (sahas3) [inline projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py]: No need to document old behavior

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir [hunk 7]
```diff
@@ -2258,8 +2258,11 @@ func.func @test_size(%arg0: !torch.vtensor<[3,4,5],f32>) -> !torch.vtensor<[],si
 
 // CHECK-LABEL: func.func @test_softplus
 func.func @test_softplus(%arg0: !torch.vtensor<[3],f32>) -> !torch.vtensor<[3],f32> attributes {torch.onnx_meta.ir_version = 3 : si64, torch.onnx_meta.opset_version = 1 : si64, torch.onnx_meta.producer_name = "backend-test", torch.onnx_meta.producer_version = ""} {
-  // CHECK: %[[EXP:.*]] = torch.aten.exp %arg0 : !torch.vtensor<[3],f32> -> !torch.vtensor<[3],f32>
-  // CHECK: torch.aten.log1p %[[EXP]] : !torch.vtensor<[3],f32> -> !torch.vtensor<[3],f32>
+  // ONNX Softplus lowers to aten.softplus (beta = 1, threshold = 20); the
+  // numerically stable expansion happens in DecomposeAtenSoftplusOp.
```
Comment (sahas3) [inline test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir]: This can be removed.

File: test/Dialect/Torch/decompose-complex-ops.mlir [hunk 8]
```diff
@@ -1666,3 +1670,31 @@ func.func @torch.aten.linalg_vector_norm$zero_dim_keepdim(%arg0: !torch.vtensor<
   %0 = torch.aten.linalg_vector_norm %arg0, %ord, %dim, %keepdim, %dtype : !torch.vtensor<[3,4],f32>, !torch.float, !torch.list<int>, !torch.bool, !torch.none -> !torch.vtensor<[3,1],f32>
   return %0 : !torch.vtensor<[3,1],f32>
 }
+
+// -----
+
+// softplus(x, beta, threshold) uses the numerically stable inner arm
+//   (max(z, 0) + log1p(exp(-|z|))) / beta,  z = x * beta
+// rather than the naive log1p(exp(z)) / beta (which overflows to +inf in fp32
```
Comment (sahas3) [inline test/Dialect/Torch/decompose-complex-ops.mlir]: Here too please remove documentation of old behavior.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (sahas3) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: So this is still valid right? The e2e test you've added now doesn't trigger the fixed path? 

In terms of testing true ONNX behavior we need to wait for https://github.com/llvm/torch-mlir/pull/4747 to land.

**Review by sahas3 — CHANGES_REQUESTED** [review]

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 5 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Done.

File: projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py (hunk 6 above)
Comment (nmalimban) [reply projects/pt1/python/torch_mlir_e2e_test/test_suite/basic.py]: Removed mention of old behavior.

File: test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir (hunk 7 above)
Comment (nmalimban) [reply test/Conversion/TorchOnnxToTorch/simple_ops_q_to_z.mlir]: Removed.

File: test/Dialect/Torch/decompose-complex-ops.mlir (hunk 8 above)
Comment (nmalimban) [reply test/Dialect/Torch/decompose-complex-ops.mlir]: Done.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: That is correct, the newly added e2e test only exercises the WHERE branch of the graph.

That being said, I thought about this more, and realized I could engineer an e2e test that does test the WHERE branch of the inner softplus decomposition. If I create a PyTorch module with a softplus op containing a high threshold (say, 100), then the exported ONNX graph will have the WHERE branch guarded by this higher threshold, allowing inputs like x = 21 to flow through to the softplus node. This will get decomposed in torch-mlir with its own WHERE branch with a threshold of 20. I can add this e2e test just to ensure that this edge case is properly handled.

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 4 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: Added

File: lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp (hunk 1 above)
Comment (nmalimban) [reply lib/Conversion/TorchOnnxToTorch/DefaultDomainQtoZ.cpp]: I have adjusted the comment above the e2e test you mention to clarify that it only hits the WHERE branch. The additional test I added has comments clarifying that the threshold is raised to test the inner WHERE branch decomposed from `aten.softplus`.

**Comment by nmalimban** [comment]
Hi @Lallapallooza @rengolin, a gentle reminder to take a look at this PR at your convenience. Thanks!

**Comment by nmalimban** [comment]
@sahas3 I made the code changes you requested. If you have no further flags, could you merge this change in? Thank you!
